src/util/stream.rs

```diff
@@ -0,0 +1,433 @@
+use std::{
+    future::Future,
+    io,
+    ops::{Deref, DerefMut},
+};
+
+use bytes::Bytes;
+use iroh::endpoint::{ReadExactError, VarInt};
+use iroh_io::AsyncStreamReader;
+use serde::{de::DeserializeOwned, Serialize};
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+
+/// An abstract `iroh::endpoint::SendStream`.
+pub trait SendStream: Send {
+    /// Send bytes to the stream. This takes a `Bytes` because iroh can directly use them.
+    fn send_bytes(&mut self, bytes: Bytes) -> impl Future<Output = io::Result<()>> + Send;
+    /// Send that sends a fixed sized buffer.
+    fn send<const L: usize>(
```

Comment:
why not make it implement `tokio::io::Write` instead?

Reply:
I tried doing a

```rust
impl<T: SendStream> tokio::io::AsyncWrite for T
```

But then you get a collision when you do the reference stuff

```rust
impl<T: SendStream> SendStream for &mut T
```

So you can either make this trait convenient to use on its own, or make it convenient to adapt to the tokio world, but not both.